### DISCUSSION TITLE
Is there a way to run coin selection on a subset of coins?

### DISCUSSION THREAD

**Discussion by icota** [bitcoindevkit/bdk#1080] [Q&A] [answered]
I wanted to run `coin_select` on `BranchAndBound` but `Utxo` is member of `types` which is private to the crate.

Use-case here is to coin select only within coins that have been selected (or previously tagged) by the user.

**Comment by notmandatory** [comment, marked as answer]
The `TxBuilder` allows you to specify which UTXOs must NOT be used in coin-selection. It's a bit awkward but if you give it all the UTXOs but the user selected subset then coin-selection will only use your selected subset.  The assumption is that you'll usually only have a few UTXOs  (such as from dusting attacks) that you want to exclude from coin-selection and the rest are fair game.

https://docs.rs/bdk/latest/bdk/wallet/tx_builder/struct.TxBuilder.html#method.add_unspendable

Going forward for new BDK APIs this sounds like a good feature to add to let you specify your allowed subset directly.  Would this make sense for `TxBuilder` functions to specify "allowed" UTXOs?

```rust
/// Add a list of outpoints to the internal list of UTXOs that are allowed to be spent. If not set then all "spendable" 
/// UTXOs are allowed.
///
/// If an error occurs while adding any of the UTXOs then none of them are added and the error is returned.
///
/// The “unspendable” UTXOs have priority over the allowed UTXOs, meaning that if a UTXO is present both in the
/// “allowed” and the “unspendable” list, it will not be spent. UTXOs added with the `add_utxo` or `add_utxos` 
/// functions will be added whether they are in the "allowed" list or not.
TxBuilder::allow_utxos(&mut self, outpoints: [OutPoint]) -> Result<&mut Self, Error>
```

```rust
/// Add an outpoint to the internal list of UTXOs that are allowed to be spent. If not set then all "spendable" 
/// UTXOs are allowed.
///
/// If an error occurs while adding the UTXO then the error is returned.
///
/// The “unspendable” UTXOs have priority over the allowed UTXOs, meaning that if a UTXO is present both in the
/// “allowed” and the “unspendable” list, it will not be spent. UTXOs added with the `add_utxo` or `add_utxos` 
/// functions will be added whether they are in the "allowed" list or not.
TxBuilder::allow_utxo(&mut self, outpoint: OutPoint) -> Result<&mut Self, Error>
```

**Comment by icota** [comment]
@notmandatory going forward it looks like a fine API indeed.

But why not allow user to do coin selection directly? User might want to tweak the results of whatever coins are returned by the algo hereself? 

In that case `build_tx` is a bit awkward, especially when you want to [not increment the address index](https://github.com/bitcoindevkit/bdk/discussions/948).

**Reply by notmandatory** [reply]
A stand-alone coin selection module is being worked on now but it's still early so no predictions yet on when it will be ready to integrate with the rest of BDK. See: https://github.com/bitcoindevkit/bdk/pull/1072
